File: packages/contracts/python/sros_contracts/registry.py
"""Registry references and entries.

Ontology V2 §14: evolving taxonomies are registry rows, not enums. This module
declares the *reference* and *entry* shapes; it never enumerates values. Doing so
would recreate the migration-per-concept problem the registry split exists to
prevent.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from enum import Enum

from .errors import ContractError
from .generated.domain import REGISTRY_NAMES, RegistryStatus

_ID = re.compile(r"^[a-z0-9][a-z0-9._-]{0,127}$")

__all__ = ["RegistryRef", "RegistryEntry", "RegistryStatus", "REGISTRY_NAMES"]


@dataclass(frozen=True, order=True)
class RegistryRef:
    """A reference into an extensible registry.

    Persists the **stable identifier**, never the display name. Storing a display
    name means a rename silently rewrites history (Ontology V2 §14.4).
    """

    registry: str
    id: str

    def __post_init__(self) -> None:
        if self.registry not in REGISTRY_NAMES:
            raise ContractError(
                "registry",
                f"unknown registry {self.registry!r}. "
                f"Closed enums (ClaimType, MarketScopeType, ResearchSessionStatus, "
                f"DemandSignalFamily, ScoreFamily) are not registries. "
                f"Known registries: {', '.join(REGISTRY_NAMES)}",
            )
        if not isinstance(self.id, str) or not _ID.match(self.id):
            raise ContractError(
                f"{self.registry}.id",
                f"registry ids are lowercase stable slugs matching {_ID.pattern}, got {self.id!r}",
            )

    def to_json(self) -> dict[str, str]:
        return {"id": self.id, "registry": self.registry}

    @classmethod
    def from_json(cls, data: object) -> RegistryRef:
        if not isinstance(data, dict):
            raise ContractError("registry_ref", "expected an object")
        unknown = set(data) - {"registry", "id"}
        if unknown:
            raise ContractError("registry_ref", f"unknown fields: {sorted(unknown)}")
        if "registry" not in data or "id" not in data:
            raise ContractError("registry_ref", "requires 'registry' and 'id'")
        return cls(registry=data["registry"], id=data["id"])


@dataclass(frozen=True)
class RegistryEntry:
    """One registry row. Ontology V2 §14.4.

    Deprecation, never deletion: a deprecated entry stops being offered for new
    classification but keeps resolving for historical records.
    """

    registry: str
    id: str
    name: str
    version: int = 1
    status: RegistryStatus = RegistryStatus.ACTIVE
    description: str | None = None
    aliases: tuple[str, ...] = field(default_factory=tuple)

    def __post_init__(self) -> None:
        RegistryRef(registry=self.registry, id=self.id)  # reuse the reference validation
        if not self.name:
            raise ContractError(f"{self.registry}.{self.id}.name", "canonical name is required")
        if not isinstance(self.version, int) or isinstance(self.version, bool) or self.version < 1:
            raise ContractError(f"{self.registry}.{self.id}.version", "version starts at 1")
        if not isinstance(self.status, RegistryStatus):
            raise ContractError(
                f"{self.registry}.{self.id}.status", "status must be a RegistryStatus"
            )

    @property
    def ref(self) -> RegistryRef:
        return RegistryRef(registry=self.registry, id=self.id)

    def to_json(self) -> dict[str, object]:
        return {
            "aliases": list(self.aliases),
            "description": self.description,
            "id": self.id,
            "name": self.name,
            "registry": self.registry,
            "status": self.status.value if isinstance(self.status, Enum) else self.status,
            "version": self.version,
        }

File: packages/contracts/src/registry.ts
/**
 * Registry references and entries.
 *
 * Ontology V2 §14: evolving taxonomies are registry rows, not enums. This
 * module declares the *reference* and *entry* shapes; it never enumerates
 * values. Doing so would recreate the migration-per-concept problem the
 * registry split exists to prevent.
 */

import { ContractError } from "./errors.ts";
import {
  REGISTRY_NAMES,
  type RegistryName,
  type RegistryStatus,
} from "./generated/domain.ts";

const ID_PATTERN = /^[a-z0-9][a-z0-9._-]{0,127}$/;

/**
 * A reference into an extensible registry.
 *
 * Persists the **stable identifier**, never the display name. Storing a display
 * name means a rename silently rewrites history (Ontology V2 §14.4).
 */
export interface RegistryRef {
  readonly registry: RegistryName;
  readonly id: string;
}

/** One registry row. Deprecation, never deletion. */
export interface RegistryEntry {
  readonly registry: RegistryName;
  readonly id: string;
  readonly name: string;
  readonly version: number;
  readonly status: RegistryStatus;
  readonly description?: string | null;
  readonly aliases?: readonly string[];
}

export function parseRegistryRef(data: unknown): RegistryRef {
  if (typeof data !== "object" || data === null || Array.isArray(data)) {
    throw new ContractError("registry_ref", "expected an object");
  }
  const record = data as Record<string, unknown>;
  const unknown = Object.keys(record).filter((k) => k !== "registry" && k !== "id");
  if (unknown.length > 0) {
    throw new ContractError("registry_ref", `unknown fields: ${unknown.sort().join(", ")}`);
  }

  const { registry, id } = record;
  if (typeof registry !== "string" || !(REGISTRY_NAMES as readonly string[]).includes(registry)) {
    throw new ContractError(
      "registry",
      `unknown registry ${JSON.stringify(registry)}. Closed enums (ClaimType, ` +
        `MarketScopeType, ResearchSessionStatus, DemandSignalFamily, ScoreFamily) ` +
        `are not registries. Known registries: ${REGISTRY_NAMES.join(", ")}`,
    );
  }
  if (typeof id !== "string" || !ID_PATTERN.test(id)) {
    throw new ContractError(
      `${registry}.id`,
      `registry ids are lowercase stable slugs matching ${ID_PATTERN.source}, got ${JSON.stringify(id)}`,
    );
  }

  return Object.freeze({ registry: registry as RegistryName, id });
}

export const registryRefToJson = (ref: RegistryRef): { id: string; registry: string } => ({
  id: ref.id,
  registry: ref.registry,
});

export function parseRegistryEntry(data: unknown): RegistryEntry {
  if (typeof data !== "object" || data === null) {
    throw new ContractError("registry_entry", "expected an object");
  }
  const record = data as Record<string, unknown>;
  const ref = parseRegistryRef({ registry: record["registry"], id: record["id"] });

  const name = record["name"];
  if (typeof name !== "string" || name.length === 0) {
    throw new ContractError(`${ref.registry}.${ref.id}.name`, "canonical name is required");
  }

  const version = record["version"] ?? 1;
  if (typeof version !== "number" || !Number.isInteger(version) || version < 1) {
    throw new ContractError(`${ref.registry}.${ref.id}.version`, "version starts at 1");
  }

  // Validate the RAW value, then narrow. Casting first would make the check
  // tautological, which is exactly what the linter caught.
  const rawStatus: unknown = record["status"] ?? "ACTIVE";
  if (rawStatus !== "ACTIVE" && rawStatus !== "DEPRECATED") {
    throw new ContractError(
      `${ref.registry}.${ref.id}.status`,
      "status must be ACTIVE or DEPRECATED",
    );
  }
  const status: RegistryStatus = rawStatus;

  const description = record["description"];
  const aliases = record["aliases"];

  return Object.freeze({
    registry: ref.registry,
    id: ref.id,
    name,
    version,
    status,
    description: typeof description === "string" ? description : null,
    aliases: Array.isArray(aliases)
      ? Object.freeze(aliases.filter((a): a is string => typeof a === "string"))
      : Object.freeze([]),
  });
}
